### PR TITLE
Feat/edit meal

### PR DESCRIPTION
**Pull request by dzmpr**
Implemented screen for creating new recipe.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -47,25 +48,30 @@ dependencies {
...
-    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.2.0"
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.3.0"
     // LiveData
-    implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.2.0'
+    implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.3.0'
     // Dagger 2
-    implementation 'com.google.dagger:dagger:2.31.2'
-    kapt 'com.google.dagger:dagger-compiler:2.31.2'
+    implementation 'com.google.dagger:dagger:2.32'
+    kapt 'com.google.dagger:dagger-compiler:2.32'
```
Comment (JorikFat) [inline app/build.gradle]: move to def

File: app/build.gradle [hunk 2]
```diff
@@ -47,25 +48,30 @@ dependencies {
...
+    implementation 'androidx.fragment:fragment-ktx:1.3.0'
     // Room
     implementation 'androidx.room:room-runtime:2.2.6'
     kapt 'androidx.room:room-compiler:2.2.6'
     // ViewModel
-    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.2.0"
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.3.0"
     // LiveData
-    implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.2.0'
+    implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.3.0'
```
Comment (JorikFat) [inline app/build.gradle]: move to def

File: app/build.gradle [hunk 3]
```diff
@@ -47,25 +48,30 @@ dependencies {
...
     // Dagger 2
-    implementation 'com.google.dagger:dagger:2.31.2'
-    kapt 'com.google.dagger:dagger-compiler:2.31.2'
+    implementation 'com.google.dagger:dagger:2.32'
+    kapt 'com.google.dagger:dagger-compiler:2.32'
+    // Navigation Component
+    implementation "androidx.navigation:navigation-fragment-ktx:2.3.3"
+    implementation "androidx.navigation:navigation-ui-ktx:2.3.3"
```
Comment (JorikFat) [inline app/build.gradle]: move to def

File: app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import pw.prsk.goodfood.databinding.FragmentAddIngredientBinding
+import pw.prsk.goodfood.utils.AutocompleteSelectionHelper
+import pw.prsk.goodfood.utils.DropdownSelectionHelper
+import pw.prsk.goodfood.utils.InputValidator
+import pw.prsk.goodfood.viewmodels.EditMealViewModel
+
+class AddIngredientBottomFragment : BottomSheetDialogFragment() {
+    private lateinit var binding: FragmentAddIngredientBinding
+
+    private val editMealViewModel: EditMealViewModel by viewModels({requireParentFragment()})
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt:27]: why ()?

File: app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    private lateinit var selectedUnitHelper: DropdownSelectionHelper
+    private lateinit var selectedProductHelper: AutocompleteSelectionHelper
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        // Open dialog with expanded state
+        val bottomSheetDialog = super.onCreateDialog(savedInstanceState) as BottomSheetDialog
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt:34]: need create from builder

File: app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        bottomSheetDialog.setOnShowListener {
+            val dialog = it as BottomSheetDialog
+            val bottomSheet =
+                dialog.findViewById<View>(com.google.android.material.R.id.design_bottom_sheet)
+            BottomSheetBehavior.from(bottomSheet!!).state = BottomSheetBehavior.STATE_EXPANDED
+        }
+        return bottomSheetDialog
+    }
+
+    override fun onCreateView(
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt:44]: redundant

File: app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        super.onCreate(savedInstanceState)
+        if (savedInstanceState == null) {
+            (requireActivity().application as MyApplication).appComponent.inject(viewModel)
+            viewModel.loadProducts()
+            viewModel.loadUnits()
+            viewModel.loadCategories()
+        }
+    }
+
+    override fun onCreateView(
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt:57]: redundant

File: app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                    binding.tilRecipeName.editText?.text.toString(),
+                    description,
+                    servingsCount,
+                    categorySelectHelper.selected as MealCategory?
+                )
+            }
+        }
+    }
+
+    private fun initIngredientsRecycler() {
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt:109]: pass binding in adapter

File: app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+            val sheet = AddIngredientBottomFragment()
+            sheet.show(childFragmentManager, null)
+        }
+
+        binding.cvRecipePhoto.setOnClickListener {
+            showPopup()
+        }
+
+        // Set default slider value to TextView
+        binding.tvServings.text = binding.sServings.value.toInt().toString()
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt:80]: why binding not in xml?

File: app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        binding.rvIngredientsList.apply {
+            adapter = listAdapter
+            layoutManager = LinearLayoutManager(this.context)
+            addItemDecoration(
+                DividerItemDecoration(this.context,
+                    LinearLayoutManager.VERTICAL)
+            )
+        }
+
+        viewModel.ingredients.observe(viewLifecycleOwner) {
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt:121]: show/hide items in adapter

File: app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            binding.groupPlaceholder.visibility = if (it.isNotEmpty()) {
+                View.GONE
+            } else {
+                View.VISIBLE
+            }
+            listAdapter.setList(it)
+        }
+    }
+
+    private fun subscribeUi() {
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/EditMealFragment.kt:131]: why need separate method?

File: app/src/main/java/pw/prsk/goodfood/adapters/MealCategoryAutocompleteAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package pw.prsk.goodfood.adapters
+
+import android.content.Context
+import androidx.annotation.LayoutRes
+import pw.prsk.goodfood.data.MealCategory
+
+class MealCategoryAutocompleteAdapter(
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/adapters/MealCategoryAutocompleteAdapter.kt:7]: redundant class

File: app/src/main/java/pw/prsk/goodfood/adapters/ProductAutocompleteAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+import android.widget.ArrayAdapter
+import android.widget.Filter
+import android.widget.Filterable
+import android.widget.TextView
+import androidx.annotation.LayoutRes
+import pw.prsk.goodfood.R
+import pw.prsk.goodfood.data.Product
+import java.util.*
+
+class ProductAutocompleteAdapter(
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/adapters/ProductAutocompleteAdapter.kt:16]: redundant class

File: app/src/main/java/pw/prsk/goodfood/adapters/ProductUnitAdapter.kt [hunk 14]
```diff
@@ -1,27 +1,13 @@
...
-import android.view.LayoutInflater
-import android.view.View
-import android.view.ViewGroup
-import android.widget.ArrayAdapter
-import android.widget.TextView
 import androidx.annotation.LayoutRes
-import pw.prsk.goodfood.R
 import pw.prsk.goodfood.data.ProductUnit
 
 class ProductUnitAdapter(
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/adapters/ProductUnitAdapter.kt]: redundant class

File: app/src/main/java/pw/prsk/goodfood/data/PhotoGateway.kt [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.FileNotFoundException
+import java.util.*
+
+class PhotoGateway(private val applicationContext: Context) {
+    fun createNewPhotoFile() : String {
+        while (true) {
+            val fileName = UUID.randomUUID().toString()
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/data/PhotoGateway.kt]: need readable file name

File: app/src/main/java/pw/prsk/goodfood/data/Converters.kt [hunk 16]
```diff
@@ -15,4 +18,15 @@ class Converters {
     fun toLocalDateTime(millis: Long): LocalDateTime {
         return LocalDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneId.systemDefault())
     }
+
+    @TypeConverter
+    fun fromIngredientsList(list: List<Ingredient>): String {
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/data/Converters.kt:23]: it must be realized via extentions

File: app/src/main/res/values/colors.xml [hunk 17]
```diff
@@ -12,7 +12,7 @@
     <color name="raisin_black">#242124</color>
     <color name="ivory">#FFFFF3</color>
 
-    <color name="light_gray">#E5E5E5</color>
+    <color name="light_gray">#E5E5E5</color> <!-- Unchecked chip background color -->
```
Comment (JorikFat) [inline app/src/main/res/values/colors.xml:15]: rename

File: app/src/main/res/menu/menu_home_navigation.xml [hunk 18]
```diff
@@ -1,17 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <menu xmlns:android="http://schemas.android.com/apk/res/android">
     <item
-        android:id="@+id/actionMeals"
+        android:id="@id/actionMeals"
```
Comment (JorikFat) [inline app/src/main/res/menu/menu_home_navigation.xml:4]: а вот за это респект

File: app/src/main/res/menu/menu_home_navigation.xml [hunk 19]
```diff
@@ -1,17 +1,17 @@
...
         android:title="@string/label_navigation_meals"
         android:icon="@drawable/ic_navigation_meals"/>
 
     <item
-        android:id="@+id/actionProducts"
+        android:id="@id/actionProducts"
```
Comment (JorikFat) [inline app/src/main/res/menu/menu_home_navigation.xml:9]: а вот за это респект

File: app/src/main/res/menu/menu_home_navigation.xml [hunk 20]
```diff
@@ -1,17 +1,17 @@
...
         android:title="@string/label_navigation_products"
         android:icon="@drawable/ic_navigation_products"/>
 
     <item
-        android:id="@+id/actionProfile"
+        android:id="@id/actionProfile"
```
Comment (JorikFat) [inline app/src/main/res/menu/menu_home_navigation.xml:14]: а вот за это респект

File: app/src/main/res/menu/menu_add_recipe_photo.xml [hunk 21]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <group android:id="@+id/groupAddPhoto">
+        <item
+            android:id="@+id/actionTakePhoto"
```
Comment (JorikFat) [inline app/src/main/res/menu/menu_add_recipe_photo.xml:7]: rename

File: app/src/main/res/menu/menu_add_recipe_photo.xml [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+            android:title="@string/take_photo_action_label" />
+        <item
+            android:id="@+id/actionChoosePhoto"
+            android:title="@string/choose_photo_action_label" />
+    </group>
+    <group android:id="@+id/groupRemovePhoto">
+        <item
+            android:id="@+id/actionRemovePhoto"
```
Comment (JorikFat) [inline app/src/main/res/menu/menu_add_recipe_photo.xml:15]: rename

File: app/src/main/res/menu/menu_add_recipe_photo.xml [hunk 23]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <group android:id="@+id/groupAddPhoto">
+        <item
+            android:id="@+id/actionTakePhoto"
+            android:title="@string/take_photo_action_label" />
+        <item
+            android:id="@+id/actionChoosePhoto"
```
Comment (JorikFat) [inline app/src/main/res/menu/menu_add_recipe_photo.xml:10]: rename

File: app/src/main/res/layout/item_ingredient.xml [hunk 24]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:orientation="horizontal"
+    android:paddingTop="9dp"
```
Comment (JorikFat) [inline app/src/main/res/layout/item_ingredient.xml]: use single tag

File: app/src/main/res/layout/item_ingredient.xml [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:id="@+id/tvIngredientName"
+        android:textSize="18sp"
+        android:textColor="?attr/colorOnSurface"
+        tools:text="Cucumber"/>
+
+    <View
```
Comment (JorikFat) [inline app/src/main/res/layout/item_ingredient.xml:18]: why this need?

File: app/src/main/res/layout/fragment_edit_meal.xml [hunk 26]
```diff
@@ -0,0 +1,265 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="?attr/colorSurface"
+    android:orientation="vertical">
+
+    <com.google.android.material.appbar.AppBarLayout
```
Comment (JorikFat) [inline app/src/main/res/layout/fragment_edit_meal.xml:10]: separate to external view

File: app/src/main/java/pw/prsk/goodfood/utils/ProductDiffUtilCallback.kt [hunk 27]
```diff
@@ -2,9 +2,8 @@ package pw.prsk.goodfood.utils
 
 import androidx.recyclerview.widget.DiffUtil
 import pw.prsk.goodfood.data.Product
-import pw.prsk.goodfood.data.ProductWithMeta
 
-class ProductDiffUtilCallback(private val oldList: List<ProductWithMeta>, private val newList: List<ProductWithMeta>) :
+class ProductDiffUtilCallback(private val oldList: List<Product>, private val newList: List<Product>) :
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/utils/ProductDiffUtilCallback.kt]: not need separate file

File: app/src/main/java/pw/prsk/goodfood/adapters/ProductAdapter.kt [hunk 28]
```diff
@@ -30,14 +29,15 @@ class ProductAdapter: RecyclerView.Adapter<ProductAdapter.ProductViewHolder>() {
     override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
         with (productList[position]) {
             holder.name.text = name
-            holder.category.text = categoryName
-            holder.unit.text = unitName
+            // TODO: Remove after fixing layout
+            holder.category.text = "deprecated"
```
Comment (JorikFat) [inline app/src/main/java/pw/prsk/goodfood/adapters/ProductAdapter.kt:32]: not fixed?

File: app/src/main/res/layout/item_ingredient.xml (hunk 25 above)
Comment (dzmpr) [reply app/src/main/res/layout/item_ingredient.xml:18]: for preview

File: app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+        val unitValidator = InputValidator(binding.tilAmountUnit, resources.getString(R.string.label_product_units_error))
+
+        selectedProductHelper = AutocompleteSelectionHelper(binding.tilIngredientName) { input ->
+            Product(name = input)
+        }.addItemSelectedListener {
+            // Set focus to next text field
+            binding.tilAmount.requestFocus()
+        }
+
+        selectedUnitHelper = DropdownSelectionHelper(binding.tilAmountUnit) { input ->
```
Comment (dmdt) [inline app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt:67]: t

File: app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt (hunk 4 above)
Comment (dzmpr) [reply app/src/main/java/pw/prsk/goodfood/AddIngredientBottomFragment.kt:27]: it isn't trailing lambda

**Pull request merged**